### PR TITLE
feat: split authorization from access

### PR DESCRIPTION
**Pull request by Gozala**
I have attempted to separate authorization protocol from access protocol

### CODE REVIEW

File: w3-access.md [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+}
+```
+
+# Examples
+
+## Access across multiple devices
+
+> Access protocol implies that delegation issuer be aware of the audience DID. Often complementary [authorization] protocol can be utilized to remove need for passing around cryptographic [`did:key`] identifiers as following example attempt to illustrate.
+
+When Alice first runs `w3up` program it asks for the user identity she'd like to use. After alice types `alice@web.mail` program initiates [authorization] protocol and creates a new (user) space by deriving `did:key:zAliceSpace` [`did:key`] from it. It immediately delegates capabilities to this space to Alice's identity:
```
Comment (Gozala) [inline w3-access.md:126]:
```suggestion
When Alice first runs `w3up` program it asks for the user identity she'd like to use. After Alice types `alice@web.mail` program initiates [authorization] protocol and creates a new (user) space by deriving `did:key:zAliceSpace` [`did:key`] from it. It immediately delegates capabilities to this space to Alice's identity:
```

File: w3-session.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+Authorization context extends to the principals that trust that notary has carried out authorization process is good faith.
+
+> To make it more concrete consider following scenario: **Alice** shares **Bobs** contact info with **Mallory** from her address book.
+>
+> **Mallory** assumes shared phone number belongs to **Bob** because she trusts **Alice** to act in good faith.
+
+#### Delegated Trust
+
+Trust in notary CAN explicitly be captured via [session] delegating, in which case notary SHOULD re-delegate it. This allows modelling complex trust relationships where trusted authority delegates trust to some intermediary principal which (re)delegates this trust to the notary that issued a [session].
```
Comment (Gozala) [inline w3-session.md:124]:
```suggestion
Trust in notary CAN explicitly be captured via [session] delegating, in which case notary SHOULD re-delegate it. This allows modeling complex trust relationships where trusted authority delegates trust to some intermediary principal which (re)delegates this trust to the notary that issued a [session].
```

File: w3-access.md [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+    "can": "*"
+  }],
+  "exp": null,
+  "sig": "..."
+}
+```
+
+## Sharing access with a friend
+
+> Access protocol could be utilized to share access with a friend who may never used web3.storage.
```
Comment (gobengo) [inline w3-access.md:224]:
```suggestion
> Access protocol could be utilized to share access with a friend who may have never used web3.storage.
```

File: w3-access.md [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+
+Alice wants to share access to her space with her friend Bob. She does not know if Bob has ever heard of the web3.storage, but she knows his email address `bob@gmail.com` allowing her to delegate capabilities to it:
```
Comment (gobengo) [inline w3-access.md:226]:
```suggestion
Alice wants to share access to her space with her friend Bob. She does not know if Bob has ever heard of web3.storage, but she knows his email address `bob@gmail.com` allowing her to delegate capabilities to it:
```

File: w3-access.md [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+        "with": "did:web:web3.storage",
+        "can": "./update",
+        "nb": { "key": "did:key:zBobAgent" }
+      }],
+    }
+  ]
+}
+```
+
+Service responds with a delegation from Alice giving Bobs agent capability to execute `store/list` on `did:key:zAliceSpace` space
```
Comment (gobengo) [inline w3-access.md:279]:
```suggestion
Service responds with a delegation from Alice giving Bob's agent capability to execute `store/list` on `did:key:zAliceSpace` space
```

File: w3-provider.md [hunk 6]
```diff
@@ -148,7 +147,7 @@ A user MAY get a "free" storage provider (from web3.storage) by invoking a self-
 
 Providers MAY impose certain requirements that resource (`with`) must meet.
 
-> Free provider (from web3.storage) requires that resource MUST be an [account][] identifier. Lite and Expert providers (from web3.storage) additionally require that the invocation resource have a payment provider (for billing purposes).
+> Free provider (from web3.storage) requires that resource MUST be a [`did:mailto`] identifier. Lite and Expert providers (from web3.storage) additionally require that the invocation resource have a payment provider (for billing purposes).
```
Comment (gobengo) [inline w3-provider.md:150]:
```suggestion
> Free provider (from web3.storage) requires that resource MUST be a [`did:mailto`] identifier. Lite and Expert providers (from web3.storage) additionally require that the invocation resource have a payment method (for billing purposes).
```

(would avoid 'payment provider' so its not confused with 'provider' in the spec itself)

File: w3-provider.md [hunk 7]
```diff
@@ -222,89 +219,88 @@ An agent MUST be delegated `consumer/add` capability, on successful [`provider/g
 
 ###### add `aud`
 
-Capability MUST be delegated back to the `iss` of the [`provider/get`][] request.
+Capability MUST be delegated back to the `iss` of the [`provider/get`] request.
 
-> This allows an [account][] to delegate `provider/get` capability to an agent or another user, which can then complete the loop using [`consumer/add`][]. If capability was delegated back to `with` identifier instead only account or delegate (with `consumer/add` capability) would be able to complete the loop.
+> This allows authorized agent to delegate `provider/get` capability to an agent or another user, which can then complete the loop using [`consumer/add`]. If capability was delegated back to `with` identifier instead only account or delegate (with `consumer/add` capability) would be able to complete the loop.
```
Comment (gobengo) [inline w3-provider.md:224]:
```suggestion
> This allows authorized agent to delegate `provider/get` capability to an agent or another user, which can then complete the loop using [`consumer/add`]. If capability was delegated back to `with` identifier instead, only account or delegate (with `consumer/add` capability) would be able to complete the loop.
```

File: w3-session.md [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+#### Authorization Key
+
+Resource (`with`) MUST be a [`did:key`][] identifier for the desired session key. It represents a key that agent wishes to use for signing [UCAN][]s that it will issue using authorization principal.
+
+> Please note that session key `with`, MAY be different or same as the issuing agent DID (`iss`).
+
+### Authentication
+
+Agent providing `access/authorize` capability (notary) MUST carry out-of-bound process in which authorizing principal may confirm or deny authorization. Process MAY vary based on type of identity.
```
Comment (gobengo) [inline w3-session.md:95]: > based on type of identity

Does this mean something different than 'based on DID method'?

File: w3-session.md [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+
+#### Example [`did:mailto`] authorization
+
+Notary MAY issue authorization [session] after verifying that specified [`did:key`] is authorized by either:
+
+- Performing [`did:mailto` resolution] and ensuring it is listed under [capability invocations] when [DKIM-Signature] is provided.
+
+- Perform out-of-bound verification by sending an email to the corresponding address with a callback link that can authorize requested [`did:key`].
+
+#### Example [`did:web`]
```
Comment (gobengo) [inline w3-session.md:105]: love to have this in here!

**Review by gobengo — APPROVED** [review]

File: w3-session.md [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+Capability provider (notary) MUST carry out-of-bound authentication that allows identity owner to accept or deny authorization request.
+
+```mermaid
+sequenceDiagram
+  participant Agent as 👩‍💻<br/><br/>did:key:zAgent
+  participant W3 as 🌐<br/><br/>did:web:web3.storage #32;
+  participant Email as 📬<br/><br/>alice@web.mail
+
+  Agent ->> W3: access/authorize
+  Note right of Agent:🎟<br/>with: did:key:zAgent<br/>as: did:mailto:alice@web.mail
```
Comment (alanshaw) [inline w3-session.md:60]: Please can we be consistent with `did:mailto:alice@web.mail` or `did:mailto:web.mail:alice`?

File: w3-session.md [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+
+## Session
+
+Session is UCAN issued by (trusted) notary (`iss`) and proof that notary has verified that authorizing principal (`aud`) has approved use of session key (`with`) in the context of this session.
+
+In other words notary has ensured that principal in `aud` field has approved UCAN issued by it to be signed using (`with`) key.
+
+### Session Example
+
+> Proof that `did:mailto:web:mail:alice` can issue UCANs signed with `did:key:zAgent` session key.
```
Comment (alanshaw) [inline w3-session.md:134]:
```suggestion
> Proof that `did:mailto:web.mail:alice` can issue UCANs signed with `did:key:zAgent` session key.
```

File: w3-access.md [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+
+Agent MAY send such a delegation(s) to the service implementing this protocol so that it's there for the audience to collect at their convenience.
+
+### `access/delegate`
+
+Agent CAN invoke `access/delegate` capability with arbitrary delegations which will be stored and made available for collection to their respective audiences.
+
+#### Example
+
+> Invoke `access/delegate` asking web3.storage to record delegation from the previous the previous example
```
Comment (alanshaw) [inline w3-access.md:56]:
```suggestion
> Invoke `access/delegate` asking web3.storage to record delegation from the previous example
```

**Review by alanshaw — APPROVED** [review]

File: w3-access.md [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+
+Field MUST be a space DID with a storage provider. Delegation will be stored just like any other DAG stored using `store/add` capability.
+
+Provider SHOULD deny service if DID in the `with` field has no storage provider.
+
+#### delegate `nb`
+
+Field is a set of delegations links represented as a JSON. Keys SHOULD be CID strings of the values to make encoding deterministic.
+
+Delegations MUST be included with an invocation. Invocations that link to delegations not included with invocation MUST be denied.
```
[the remark this answers is not in the document]
Comment (Gozala) [reply w3-access.md:92]: It is currently the only convenient way to do it in ucanto. Linked issue will lift this limitation and would basically allow you to stick delegations in `nb` filed.

That said it is possible to include delegations by manually adding them to the CAR that is send to the service, you could do that in ucanto as well, but it's quite messy. Other clients that don't use ucanto also might have ways to include blocks in the CARs they send.

I did not want to guide spec design by ucanto limitations, which is why it does not say put it in proofs and rather has a warning below pointing out the limitation and a workaround.

File: w3-access.md [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+}
+```
+
+# Examples
+
+## Access across multiple devices
+
+> Access protocol implies that delegation issuer be aware of the audience DID. Often complementary [authorization] protocol can be utilized to remove need for passing around cryptographic [`did:key`] identifiers as following example attempt to illustrate.
+
+When Alice first runs `w3up` program it asks for the user identity she'd like to use. After Alice types `alice@web.mail` program initiates [authorization] protocol and creates a new (user) space by deriving `did:key:zAliceSpace` [`did:key`] from it. It immediately delegates capabilities to this space to Alice's identity:
```
[the remark this answers is not in the document]
Comment (Gozala) [reply w3-access.md:126]:
```suggestion
When Alice first runs `w3up` program it asks for the user identity she'd like to use. After Alice types `alice@web.mail` program initiates [authorization] protocol and creates a new (user) space by deriving `did:key:zAliceSpace` [`did:key`] from it. It immediately delegates capabilities to this space to an identifier controlled by Alice's email address:
```

I don't mind rephrasing per your suggestion, but I while I do realize controversy here I do think that delegation is to Alice's identity encoded as did:mailto identifier.

While DIDs really try to be identifiers, we're here really are mapping did:key identifiers to user identities even if that's not what DIDs had been designed for. Which is another reason why I feel like just mailto URI would have been a better choice.

File: w3-provider.md (hunk 6 above)
Comment (Gozala) [reply w3-provider.md:150]: > would avoid 'payment provider' so its not confused with 'provider' in the spec itself

The thing is I actually meant that space should have "payment provider"

File: w3-session.md (hunk 10 above)
Comment (Gozala) [inline w3-session.md:60]:
```suggestion
  Note right of Agent:🎟<br/>with: did:key:zAgent<br/>as: did:mailto:web.mail:alice
```

Comment (Gozala) [reply w3-session.md:60]:
```suggestion
  Note right of Agent:🎟<br/>with: did:key:zAgent<br/>as: did:mailto:web.mail:alice
```
Good catch, I have missed to update it here.

File: w3-session.md (hunk 8 above)
Comment (Gozala) [reply w3-session.md:95]: Probably no, but perhaps there will be a case. I don't think it's worth committing to that exactly.

File: w3-session.md [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+Notary MAY issue authorization [session] after performing [DID resolution] and either:
+
+- Ensuring that specified [`did:key`] is authorized to perform [capability invocations].
+- Carry [DID authentication] using challenge-response protocol.
+
+### Authorization context
+
+Notary MUST issue [session] for the [authorization] request that was approved. This enables authorized agents (ones with access to session private key) to use UCANs covered by this [session].
+
+Authorization context extends to the principals that trust that notary has carried out authorization process is good faith.
```
Comment (Gozala) [inline w3-session.md:116]:
```suggestion
Authorization context extends to the principals that trust that notary has carried out authorization process in good faith.
```

File: w3-session.md [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+
+## Session Scope
+
+Authorization [session] covers UCANs that include it in their [proofs] and are subject to UCAN [time bounds][] and [revocations][ucan revocation]. Only UCANs meeting standard [principal alignment] requirement are covered by the [session], meaning UCAN MUST be issued by the `aud` of the included [session].
+
+## Session Extensions
+
+The `./update` capability provides functionality roughly equivalent of [`Set-Cookie`][] HTTP header. It allows capturing stateful information in the stateless UCAN delegations.
+
+Capability MAY include arbitrary session information besides `key` which MUST ignored in the context of this specification, or put it differently have no how authorization [session] handling.
```
Comment (Gozala) [inline w3-session.md:174]:
```suggestion
Capability MAY include arbitrary session information besides `key` which MUST be ignored in the context of this specification, or put it differently MUST NOT impact authorization [session] handling.
```

**Review by hugomrdias — APPROVED** [review]

**Pull request merged**